app/models/manageiq/providers/nuage/inventory/collector/target_collection.rb

```diff
@@ -173,7 +173,15 @@ def infer_related_ems_refs!
   end
 
   def infer_related_ems_refs_db!
-    # infer related entities based on VMDB here if needed
+    # Delete related security groups upon router deletion.
+    # TODO(miha-plesko): when we have relation "security group belongs to network router" modelled, we can use
+    # dependent: destroy instead inferring here.
+    references_with_options(:network_routers).each do |router|
+      next unless router[:operation] == 'DELETE'
+      if (router = manager.network_routers.find_by(:ems_ref => router[:ems_ref]))
```

Comment:
I think we can do this via 1 query

Something like:

```
manager.security_groups.joins(:cloud_tenant => :network_routers).where({:cloud_tenant => {:network_routers => {:ems_ref => all_router_ems_refs_array}}})
```

Reply:
Nice! Let me apply, but since it looks like the schema PR will be taken in, I hope we can omit this query completely. Because the expensive part is that then upon targeted refresh we will be sending one API request per SecurityGroup for all SecurityGroups in entire CloudTenant - while in fact we know upfront that only those related to this network router should be removed. Zero API calls needed, just removing from VMDB 🎆 🍾 😄